master/internal/kubernetes/queue.go

```diff
@@ -0,0 +1,215 @@
+//       requests.
+//
+//  When requests come in they are buffered by the requestQueue until a worker becomes available
+//  at which point the longest queue request is forwarded to the available. Requests are buffered
+//  rather than forward right away because buffering makes it possible to cancel creation requests
+//  after they are created, but before they are executed. Since the actor system processes messages
+//  in a FIFO order, if all request were forwarded right away any cancellation request would only
+//  be processed after the creation request case already been processed, requiring an unnecessary
+//  resource creation and deletion. An example of this is when a large HP search is created and
+//  then killed moments later. By having requests be buffered, if a a deletion request arrives
+//  prior to the creation request being executed, the requestQueue detects this and skips the
+//  unnecessary creation / deletion.
+//
+//  The message protocol consists of `createKubernetesResources` and `deleteKubernetesResources`
+//  messages being sent to the requestQueue. If it forwards the request to a worker, the worker
+//  will send the original task handler a `resourceCreationFailed` or a `resourceDeletionFailed`
+//  if an error was encountered while creating / deleting the resources. If a deletion request
+//  arrives before the creation request had been sent to the worker, the `requestQueue` will
+//  notify the task handler of this by sending a `resourceCreationCancelled` message.
+//  requestProcessingWorkers notify the requestQueue that they are available to receive work
+//  by sending a `workerAvailable` message.
+
+type requestQueue struct {
+	podInterface       typedV1.PodInterface
+	configMapInterface typedV1.ConfigMapInterface
+
+	queue                    []*queuedResourceRequest
+	pendingResourceCreations map[*actor.Ref]*queuedResourceRequest
+	availableWorkers         []*actor.Ref
+}
+
+func newRequestQueue(
+	podInterface typedV1.PodInterface,
+	configMapInterface typedV1.ConfigMapInterface,
+) *requestQueue {
+	return &requestQueue{
+		podInterface:       podInterface,
+		configMapInterface: configMapInterface,
+
+		queue:                    make([]*queuedResourceRequest, 0),
+		pendingResourceCreations: make(map[*actor.Ref]*queuedResourceRequest),
+		availableWorkers:         make([]*actor.Ref, 0, numKubernetesWorkers),
+	}
+}
+
+func (r *requestQueue) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+		for i := 0; i < numKubernetesWorkers; i++ {
+			newWorker, ok := ctx.ActorOf(
+				fmt.Sprintf("kubernetes-worker-%d", i),
+				&requestProcessingWorker{
+					podInterface:       r.podInterface,
+					configMapInterface: r.configMapInterface,
+				},
+			)
+			if !ok {
+				return errors.Errorf("%s actor already exists", newWorker.Address())
+			}
+		}
+
+	case createKubernetesResources:
+		if err := r.receiveCreateKubernetesResources(ctx, msg); err != nil {
```

Comment:
blocking: we probably shouldn't interrupt the whole actor by return the error. We probably should just log it.

Reply:
Yeah we should really never hit this condition anyways, so its hard to reason if the actor will be able to function correctly if we do end up hitting it. I will go ahead and remove it and just add a log message instead.